data_structures/linked_list/deque_doubly.py

```diff
@@ -0,0 +1,144 @@
+        successor = node._next
+        
+        predecessor._next = successor
+        successor._prev = predecessor
+        self._size -= 1
+        temp = node._data
+        node._prev = node._next = node._data = None
+        del node
+        return temp
+
+class LinkedDeque(_DoublyLinkedBase):
+    
+    def first(self):
+        """ return first element 
+        >>> d = LinkedDeque()
+        >>> d.add_first('A') # doctest: +ELLIPSIS
+        <deque_doubly.LinkedDeque object at ...
+        >>> d.first()
+        'A'
+        >>> d.add_first('B') # doctest: +ELLIPSIS
+        <deque_doubly.LinkedDeque object at ...
+        >>> d.first()
+        'B'
+        """ 
+        if self.is_empty():
+            raise Exception('List is empty')
+        return self._header._next._data
+    
+    def last(self):
+        """ return last element
+        >>> d = LinkedDeque()
+        >>> d.last()
+        Traceback (most recent call last):
+           ...
+        Exception: List is empty
+        >>> d.add_last('A') # doctest: +ELLIPSIS
+        <deque_doubly.LinkedDeque object at ...
+        >>> d.last()
+        'A'
+        >>> d.add_last('B') # doctest: +ELLIPSIS
+        <deque_doubly.LinkedDeque object at ...
+        >>> d.last()
+        'B'
+        """
+        if self.is_empty():
+            raise Exception('List is empty')
+        return self._trailer._prev._data
+    
+    ### DEque Insert Operations (At the front, At the end) ###
+    
+    def add_first(self, element):
+        """ insertion in the front
+        >>> d = LinkedDeque()
+        >>> d.add_first('AV') # doctest: +ELLIPSIS
+        <deque_doubly.LinkedDeque object at ...
+        """
+        return self._insert(self._header, element, self._header._next)
+    
+    def add_last(self, element):
+        """ insertion in the end
+        >>> d = LinkedDeque()
+        >>> d.add_last('A') # doctest: +ELLIPSIS
+        <deque_doubly.LinkedDeque object at ...
+        """
+        return self._insert(self._trailer._prev, element, self._trailer)
+    
+    ### DEqueu Remove Operations (At the front, At the end) ###
+    
+    def remove_first(self):
+        """ removal from the front
+        >>> d = LinkedDeque()
+        >>> d.remove_first()
+        Traceback (most recent call last):
+           ...
+        Exception: List is empty
+        >>> d.add_first('A') # doctest: +ELLIPSIS
+        <deque_doubly.LinkedDeque object at ...
+        >>> d.remove_first()
+        'A'
+        """
+        if self.is_empty():
+            raise Exception('List is empty')
+        return self._delete(self._header._next)
+    
+    def remove_last(self):
+        """ removal in the end
+        >>> d = LinkedDeque()
+        >>> d.remove_last()
+        Traceback (most recent call last):
+           ...
+        Exception: List is empty
+        >>> d.add_first('A') # doctest: +ELLIPSIS
+        <deque_doubly.LinkedDeque object at ...
+        >>> d.remove_last()
+        'A'
```

Comment:
Please add a test to make sure that d is empty.

Reply:
Yeah, i've done something like ... 
```
>>> d = LinkedDeque()
>>> d.is_empty()
True
>>> d.add_first('A') # doctest: +ELLIPSIS
<deque_doubly.LinkedDeque object at ...
>>> d.remove_last()
 'A'
>>> d.is_empty()
True
```
is it okay sir?

Reply:
```
<deque_doubly.LinkedDeque object at ...
—>
<linked_list.deque_doubly.LinkedDeque object at ...
```

Reply:
I don't know why you added those two lines after <deque_doubly ... 
are you running this from the linked_list directory?